### ISSUE TITLE
Docker size

### ISSUE THREAD

**Issue by michaelarnauts** [home-assistant/architecture#103]
I would like to open the discussion on the docker image, since I'm a bit worried about the growing size of the docker image, and the policy we have for adding things to the docker image.

Recently, tensorflow was added to the docker image (https://github.com/home-assistant/home-assistant/pull/18191 and https://github.com/home-assistant/home-assistant/pull/17795#issuecomment-435331879) causing it to grow with 400MB compared to 0.81.

```
0.82:
homeassistant/home-assistant      latest              5f6a96c200fc        2 hours ago         2.2GB

0.81:
homeassistant/home-assistant      <none>              b3d8aa8919a7        6 days ago          1.79GB
```

Tensorflow folder in the docker image:
```
root@ec5a707d8ce2:/usr/local/lib/python3.6/site-packages/tensorflow# du -sh .
362M    .
root@ec5a707d8ce2:/usr/local/lib/python3.6/site-packages/tensorflow#
```
Dockerhub compresses the layers, but it's still a 100MB extra download. (https://hub.docker.com/r/homeassistant/home-assistant/tags/)

Now, in my opinion, the official Home Assistant docker image should contain everything to make Home Assistant work (read: python libraries), but not all the binaries needed for all the components. This can lead to several gigabytes. I'm thinking about opencv, tensorflow, ssocr here.

Since hass.io is our "customer-friendly" way of running home assistant, I think those dependencies should be handled as hass.io addons, with a different docker image. 

This way, the official docker image can be kept as small as possible (we are still aiming on rpi users) and users needing object recognition and other fancy new stuff can just run the hass.io addon.

**Comment by dgomes** [comment]
I support this view...

In the past it was also referred to me that we should not even add large python requirements such as numpy or scipy.

**Comment by frenck** [comment]
It has been added to Hass.io as well: <https://github.com/home-assistant/hassio-homeassistant/pull/15/files>

Especially on Raspberry Pi's... I think this is a really bad idea.

**Comment by pvizeli** [comment]
Well, we removed it four days ago and added this into an Add-on.

This problem is what we need challenging with Docker, and therefore it exists not a way. The Add-on solution with local dept folder is not practicable and currently only a workaround.

I think the end solution is what @frenk does with community add-ons and extended packages on startup.

Anyway, all solution will be at the end Hass.io  and no docker solution because it needs extended logic and a system they can manage this for users.

**Comment by hunterjm** [comment]
I can see how we might need a policy to prevent excess bloat.  At the same time, storage is relatively cheap.

Home Assistant, when installed via the venv, actually does not install all dependancies out of the gate.  Both Hass.io and the Docker image do, however.  I am sure this was done because pre-installing and bundling with the docker image takes a lot less time for the user than trying to have the component install on demand.

I can see both sides of this argument.  Wherever this ends up, I think simplicity is more important than size for the average user.  If a power user want a slimmed down version, it is pretty easy to build a container that doesn't pre-install any of the dependancies.

I agree with @pvizeli that moving it out of the base image on Hass.io is fine, but a shared deps folder is a work-around.  You can't do that with the official docker image though.  There is no shared "add-on", so you would have to go into the shell of the official image to install additional deps every time you pulled a new version, which again is prioritizing image size over ease-of-use.

@frenck - Why do you think having it on a Pi is a bad idea?  The framework is being used in a ton of projects for the Pi, including [self driving RC cars](http://www.donkeycar.com/)

**Comment by michaelarnauts** [comment]
@hunterjm I think what @frenck meant is the general practice of adding large dependencies to the docker image, not specifically to tensorflow.

In general, i see the home assistant image as the one for "power users" that install the image themselves, so they know how docker works and are able to add additional containers if needed. the hass.io image makes it all simple and allows adding additional add-ons.

**Comment by frenck** [comment]
Well in case of Tensorflow, this can be an external container (add-on or just Docker), which solves a large part probably.

But there is a general issue to discuss/solve here @michaelarnauts, which is not just Tensorflow. With the current development speed of Home Assistant, the images will keep growing. This requires a different approach in order to handle this growth. Tensorflow is just the one triggering it now.

**Comment by hunterjm** [comment]
@frenck - I agree.  The balance to strike is size vs simplicity.  The add-on will work for Hass.io since Home Assistant uses a local deps folder that the add-on can write to.  Something similar can be done with the base docker image by having a secondary image to pull that mounts the same deps directory to install the wheel into, but that is not close to user friendly.

To keep the images smaller, the practice of auto-installing `requirements_all.txt` can be stopped as well.  If the docker images do not automatically install `requirements_all.txt`, and let the components hot load the requirements, large libraries like tensorflow can be added back to the `REQUIREMENTS` list, and the add-ons can be used to satisfy the requirement for distributions that do not have a pre-compiled wheel.  At that point, since the base image is on Ubuntu, a wheel would not be required for the HA docker image.

**Comment by dmwyatt** [comment]
I do not think the version of tensorflow in the official docker image even supports the raspberry pi anyway.

I just tried it on a pre-Sandy Bridge i5 and it tells me that TensorFlow was compiled to use AVX, but those aren't available on this particular CPU.

I don't have it on a Pi right now, but I do know the Pi does not support AVX either....

**Comment by danieledwardgeorgehitchcock** [comment]
Instead of a blanket removal of 'bloat' dependencies, why not use the power of Docker Hub to allow the user to choose what they want?

Builds can be configured to publish to tags in Docker Hub - Why not leave dependencies in for the 'latest' Docker build so that the average user will have a fully working HA but also have a 'lite' tag for stock Home Assistant without dependencies for the more advanced users?

**Comment by pvizeli** [comment]
Because you need a build farms to deploy all this architectures an a time where the user need it. But I'm open for contributions to a build system they can run distributed and provide this functionality.

Next problem is. You want component X and component Y. But they spited now of two docker images. Or you want vlc and need now use the extend docker images of 1Gb for only a binary of 40mb.

The only way to be flexible is to have a slim images and allow do install additional apks/wheel on container build.

**Comment by michaelarnauts** [comment]
Dockerhub does exactly that. It builds the images for you. CPU time is sponsored by Docker I guess.

But I prefer the suggestion of @pvizeli , a slim image, that can install dependencies on demand. 

You could use a volume to store those downloaded extra's outside the container (on the host), so you don't need to download them again when you update or recreate your container.

**Comment by OttoWinter** [comment]
Just a few comments from me:

 * Yes, this is very much a problem. I also want to add here that the huge docker image size is also really bad for users with a rather slow internet connection. Downloading home assistant updates through docker (and through HassIO) just takes *ages*.
 * I like the idea of having pre-packaged wheels. I guess it would kind of replicate the layering feature of docker images.
 * I do not like the idea of having different docker image variants due to several reasons.
    A) we would need to decide a list of integrations that are commonly used for most people. Deciding what belongs where and maintaining that would be a nightmare.
    B) This would make home assistant less user friendly for testing out new stuff. Right now, if I want to add a new integration, I just add it into the YAML and hass takes care of installing everything. Otherwise I would need to look up what image to use.
    C) these variants would almost certainly not work for many people. If there's only *a single* integration that is not in a given variant, I instantly have to resort to a bigger one or even the "full" image again.
    D) Though CPU time on docker hub *is* free, from what I understand that is currently only possible for the amd64 architecture. Hence all the HassIO builds are also not done using docker hub.
 * With community add-ons we can not achieve everything. The problem is that a real "component" in hass has tight integration with the Hass state machine, events, etc. - HassIO currently does not have that (or only through a very complicated way using the HTTP API)
 * Either way, most of these features depend on something living on the **host system** to install additional packages on-demand. Thus, this will only work for HassIO and not for generic docker installs (which is ok).

**Comment by pvizeli** [comment]
Yes, docker build system is only for amd64(i386). It's possible to use cross-build with qemu but to build a full armhf image that takes 12-16 hours. In the free variant, you can also only have one build at the time. We can invest the money into a docker enterprise account, but that solves not the issue with armhf/aarch64 and future architectures.

As next, we build for every IoT device an own image, based on the release image build that is basing on the base image. That allows us to cache several layers. So we build for every release 16 images.

For allow to customize on container layer, I see no issue with docker layering. The container layer is created to do things like that. So if we can have a base that supports 95% - 98% of components with 200mb, that should be fine. Also if we can cache the first 150mb and the user need only download around 50-60mb per release update.

**Comment by michaelarnauts** [comment]
@balloob since you've approved the tensorflow inclusion, what is your view on this?

I guess this should be the exact reason why hass.io exists, so people can install those addon-software easily, and why people should want to use that instead of a normal installation of hass.

**Comment by balloob** [comment]
I approved Tensorflow just to temporarily unblock the release. I do hope that we can get to a better solution . As you can already see these closed PRs lining up referring this issue, it's something that we should decide on a course and stick to it.

I also think that we should limit this discussion to Home Assistant and not Hass.io.

Maybe we should provide two images, home-assistant and home-assistant-build, the second one containing the build tools so people can build their own stuff. Any package that is over 1MB will have to be installed on demand. Some of those might require the `home-assistant-build` image.

**Comment by balloob** [comment]
And yes, maybe some things will end up not working on either of our images, and I think that is fine. If we try to integrate with every service and device in the world out there, it cannot be expected from us to create a container that is compatible with everything.

**Comment by d1slact0r** [comment]
I do not have much knowledge of what HA needs for python modules. but I saw that the docker file uses python 3.6 uses
```
FROM python:3.6
```
Isn't it possible to use either:
```
FROM python:3.6-slim
```
or the more secure:
```
FROM python:3.7.2-slim
```
Or even a lot smaller (which I prefer!)
```
FROM python:3.7.2-alpine3.8
```
Shrinks down the docker a lot.

**Comment by Jc2k** [comment]
FWIW:

I normally use `alpine:3.8` directly for my python projects, but in HA's case i know of a few modules it would actually break. E.g. there is an ongoing issue with the version of OpenSSL in alpine being too old to support X25519 which is needed by things using the latest release of aiossh (so breaks auswrt). Also means homekit_controller doesn't work, too.

I guess the build times would also be slower as it wouldn't be able to use wheels.

**Comment by michaelarnauts** [comment]
Also, I've had cases where musl (alpine) was a lot slower versus libc (normal version). We also depend in some debian packages that might not be in the alpine repos.

In this case, I would prefer to stick to the non-alpine version, but the python-slim variant could be interesting. Although this issue is really to discuss how we want to handle large dependencies for specific integrations or features.

**Comment by pvizeli** [comment]
musl is with GCC 8 not so slow as with GCC 4 (old alpine).

We stick on alpine with Hass.io because they support a lot more CPU architecture and musl perform on some hardware better as glibc.

**Comment by benleb** [comment]
Some experience from me... I experimentally build an image with HA on alpine. Worked without problems with the components I use 👍 Until the day it seemed to stuck in an endless loop sending iOS push messages (worked a few days without issues before). I do not know of this was caused by alpine in a way but since then I switched back to the debian image...

**Comment by michaelarnauts** [comment]
@pvizeli okay, that's nice! My experience was from over a year ago, so it might have changed since then. And since hass.io containers are alpine, that means that it should work fine for most users (I assume hass.io has a large userbase compared to the normal docker images?)

**Comment by scstraus** [comment]
Hi all, 

Was there ever any consensus what to do about about a Tensorflow Addon? To me, the obvious solution for hass.io is to run tensorflow in a docker container and just use a gRPC stub (or perhaps REST interface) in the hass.io container to access it. The stub would be very small so wouldn't cause size concerns.

See [here](https://hub.docker.com/r/tensorflow/tf_grpc_server)
and [here](https://medium.com/@KailaGaurav/deploying-object-detection-model-with-tensorflow-serving-part-3-6a3d59c1e7c0)

Theoretically, different docker containers could be created for GPU's etc (indeed the official docker contaners already come in GPU and non GPU variants). 

I'd really like to see tensorflow in Hass.io, so would like to see some progress on this. Happy to help if I can.

**Comment by maxandersen** [comment]
I thought of the same thing as I have NAS i Can run tensor flow on.

**Comment by maxandersen** [comment]
@scstraus I think if you can make a PR to enable grpc for the tensor flow component, maybe as an option in case it has overhead then that would be a great start!

**Comment by scstraus** [comment]
I certainly wouldn't mind doing that, but I don't think I'd be able to write the integration properly to homeassistant as I don't understand the internals well enough (I have looked at writing components in the past and while actually getting some code to do what I want wasn't difficult, getting it to fit into the hass architecture was a bit out of my league). Would it do to open a PR just with some sample code?

**Comment by maxandersen** [comment]
@scstraus since we already have tensorflow to integrated you wouldn't need to "Fit into the hass architecture" since thats already done. Take a look at https://github.com/home-assistant/home-assistant/blob/master/homeassistant/components/image_processing/tensorflow.py and look for tensorflow references and try replace it with the corresponding grpc variation and that would be a great start. 

I suggest you open a new PR for this and we can work through it there. even if it doesn't make it in on first go it would be an interesting experiment IMO.

**Comment by scstraus** [comment]
Good point. It doesn't look as hard as I was expecting, no PyPi component which helps, one less thing to learn. I have a clean hass.io install on Debian server that I can use. I'll start setting up the dockers for a test environment and see how far I can get in creating a gRPC version. From what I can understand from [here](https://github.com/home-assistant/home-assistant/blob/dev/CONTRIBUTING.md), I shouldn't open a PR until I have something working, so it seems I should hold off on that for now? Maybe I should work from a fork until then?

**Comment by maxandersen** [comment]
To do a PR you need a fork anyway :) if you get something moving post a url and I’ll happily help/try out what I can.

**Comment by scstraus** [comment]
I've started a thread to discuss my progress (or lack thereof). If anyone feels like following along and giving (much appreciated) advice in the hopes of dockerizing a tensorflow component, head over here:

https://community.home-assistant.io/t/attempting-to-make-a-tensorflow-hass-io-addon-by-implementing-the-tensorflow-module-with-grpc-support/99453

**Comment by frenck** [comment]
This architecture issue is rather old and still open :)

In the meantime, we have migrated to a single Docker structure for all installation types that use Docker (including the previously called Hass.io). 

Images are now based on Alpine, and have been reduced in size. More optimizations for that are on the roadmap but are out of scope of the original issue presented here.

I consider therefore this issue solved, for now, especially considering it was based on an image that we don't have anymore 😉

**Comment by tjunussov** [comment]
What about HACS  style approach ? load components as HACS does

**Comment by kbalint** [comment]
Hi, I just recently checked back again to this issue, and found the following "new" items in the 2.7 Gb docker Image (2026.8.2)

| Module | Size | Required? | Notes |
|--------|-------|-----------|-------|
| knx_frontend | 134 MB | ❌ No | KNX protocol UI; unnecessary unless using KNX |
| insteon_frontend | 71 MB | ❌ No | Insteon protocol UI |
| googleapiclient | 75 MB | ❌ No | Google API SDK (Drive, Photos, Calendar, Assistant, Cloud) |
| google | 18.9 MB | ❌ No | Google Python SDK components |
| openai | 18.7 MB | ❌ No | HA AI Conversation / LLM integration |
| selenium | 22.2 MB | ❌ No | Browser automation; legacy testing module |
| speech_recognition | 42.6 MB | ❌ No | Offline speech recognition |
| scipy | 83.5 MB | ❌ No | Advanced audio/video and mathematical processing |
| debugpy | 27.3 MB | ❌ No | Python debugger |

I have never ever heared about knx neither insteon, and don't use any google or openai proprietary stuff (are they sponsoring anything here?);  never want to use selenium, never want to have speech recognition or audio/video processing in HA, and further more I never want to debug on my live install :)

As I see, knx is used in 1.6% of installs, insteon is a ~1000 userbase. I really liked the shift when tensorflow been removed, can these packages be made optional as well? We could save some $$ on the bandwidth and disk. Thanks!
